lib/propolis/src/vmm/time_adjust.rs

```diff
@@ -0,0 +1,348 @@
+pub struct HostTime {
+    pub hrtime: Duration,
+    pub wall_clock: Duration,
+}
+
+/// Errors related to making timing adjustment calcultions
+#[derive(Clone, Debug, Error)]
+pub enum TimeAdjustError {
+    /// Error reading system clock
+    #[error("could not read system wall clock: {0}")]
+    RealTimeClock(String),
+
+    /// Error reading high res clock
+    #[error("could not read system monotonic clock: {0}")]
+    HighResClock(String),
+
+    /// Error calculating migration time delta
+    #[error("invalid migration delta: src={src_wc:?},dst={dst_wc:?}")]
+    InvalidMigrateDelta {
+        /// source host wall clock time
+        src_wc: Duration,
+
+        /// destination host wall clock time
+        dst_wc: Duration,
+    },
+
+    /// Error calculating guest uptime
+    #[error(
+        "guest uptime cannot be represented: \
+        src_hrtime={src_hrt:?}, boot_hrtime={boot_hrtime}"
+    )]
+    GuestUptimeOverflow {
+        /// source host hrtime
+        src_hrt: i128,
+
+        /// input guest boot_hrtime
+        boot_hrtime: i128,
+    },
+
+    /// Invalid total delta for boot_hrtime calculations
+    #[error(
+        "could not calculate time delta:\
+            guest uptime {uptime:?}, migrate_delta={migrate_delta:?}"
+    )]
+    TimeDeltaOverflow {
+        /// guest uptime
+        uptime: Duration,
+
+        /// migration time delta
+        migrate_delta: Duration,
+    },
+
+    /// Invalid calculated boot_hrtime
+    #[error(
+        "guest boot_hrtime cannot be represented:\
+            desc=\"{desc}\", total_delta={total_delta:?},\
+            boot_hrtime={boot_hrtime}, dst_hrtime={dst_hrtime:?}"
+    )]
+    BootHrtimeOverflow {
+        /// error description
+        desc: String,
+
+        /// calculated total delta (uptime + migration delta)
+        total_delta: Duration,
+
+        /// input guest boot_hrtime
+        boot_hrtime: Hrtime,
+
+        /// destination host hrtime
+        dst_hrtime: Duration,
+    },
+
+    /// Invalid guest TSC adjustment
+    #[error(
+        "could not calculate TSC adjustment:\
+            desc=\"{desc:?}\", migrate_delta={migrate_delta:?},
+            guest_hz={guest_hz}, tsc_adjust={tsc_adjust}"
+    )]
+    TscAdjustOverflow {
+        /// error description
+        desc: String,
+
+        /// migration time delta
+        migrate_delta: Duration,
+
+        /// guest TSC frequency (hz)
+        guest_hz: u64,
+
+        /// calculated TSC adjustment
+        tsc_adjust: u128,
+    },
+
+    /// Guest TSC overflow
+    #[error("guest TSC overflow: old tsc = {tsc}, adjustment = {adjust}")]
+    GuestTscOverflow {
+        /// input guest TSC
+        tsc: u64,
+
+        /// calculated TSC adjustment
+        adjust: u64,
+    },
+}
+
+/// Returns a non-atomic snapshot of the current hrtime and wall clock time.
+// TODO: We may want to put a bound in here to make sure the two
+// values are close enough, and we may want that to be tunable. What's the
```

Comment:
To clarify: close enough to what? IIUC these two values aren't directly comparable (since CLOCK_REALTIME uses the epoch as a base but CLOCK_HIGHRES has an arbitrary base), though I suppose you could do a high-low-high kind of thing where you read one, then read the other, then read the first one again and make sure the delta from the first reading is less than X (and you can retry if it's not).

Reply:
reworked this to just do another read of the VMM_TIME data (which exposes the wall clock / hrtime) in 803081